proposals/stack-switching/Explainer.md

```diff
@@ -0,0 +1,1098 @@
+    - and `C.tags[$e] : tag $ft`
+    - and `C.types[$ft] ~~ func [te*] -> []`
+    - and `(hdl : t2*)*`
+
+- `hdl = (on <tagidx> <labelidx>) | (on <tagidx> switch)`
+  - Handlers attached to `resume` and `resume_throw`, handling control tags for `suspend` and `switch`, respectively.
+  - `(on $e $l) : t*`
+    - iff `C.tags[$e] = tag $ft`
+    - and `C.types[$ft] ~~ func [t1*] -> [t2*]`
+    - and `C.labels[$l] = [t1'* (ref null? $ct)]`
+    - and `t1* <: t1'*`
+    - and `C.types[$ct] ~~ cont [t2'*] -> [t'*]`
+    - and `[t2*] -> [t*] <: [t2'*] -> [t'*]`
+  - `(on $e switch) : t*`
+    - iff `C.tags[$e] = tag $ft`
+    - and `C.types[$ft] ~~ func [] -> [t*]`
+
+- `suspend <tagidx>`
+  - Use a control tag to suspend the current computation.
+  - `suspend $t : [t1*] -> [t2*]`
+    - iff `C.tags[$t] = tag $ft`
+    - and `C.types[$ft] ~~ func [t1*] -> [t2*]`
+
+<!-- - `barrier <typeidx <typeidx> instr* end` -->
+<!--   - Prevents suspensions propagating beyond this program point. -->
+<!--   - `barrier $l $ft instr* end : t2*` -->
+<!--     - iff `C.labels[$l] = [t2*]` -->
+<!--     - and `C.types[$ft] ~~ func [t1*] -> [t2*]` -->
+<!--     - and `instr* : t2*` -->
+
+- `switch <typeidx> <tagidx>`
+  - Switch to executing a given continuation directly, suspending the current execution.
+  - The suspension and switch are performed from the perspective of a parent `(on $e switch)` handler, determined by the annotated control tag.
+  - `switch $ct1 $e : [t1* (ref null $ct1)] -> [t2*]`
+    - iff `C.tags[$e] = tag $ft`
+    - and `C.types[$ft] ~~ func [] -> [t*]`
+    - and `C.types[$ct1] ~~ cont [t1* (ref null? $ct2)] -> [te1*]`
+    - and `te1* <: t*`
+    - and `C.types[$ct2] ~~ cont [t2*] -> [te2*]`
+    - and `t* <: te2*`
+
+### Execution
+
+The same control tag may be used simultaneously by `throw`, `suspend`,
+`switch`, and their associated handlers. When searching for a handler
+for an event, only handlers for the matching kind of event are
+considered, e.g. only `(on $e $l)` handlers can handle `suspend`
+events and only `(on $e switch)` handlers can handle `switch`
+events. The handler search continues past handlers for the wrong kind
+of event, even if they use the correct tag.
+
+### Binary format
+
+We extend the binary format of composite types, heap types, and instructions.
+
+#### Composite types
+
+| Opcode | Type            | Parameters | Note |
+| ------ | --------------- | ---------- |------|
+| -0x20  | `func t1* t2*`  | `t1* : vec(valtype)` `t2* : vec(valtype)` | from Wasm 1.0 |
+| -0x23  | `cont $ft`      | `$ft : typeidx` | new |
+
+#### Heap Types
+
+The opcode for heap types is encoded as an `s33`.
+
+| Opcode | Type            | Parameters | Note |
+| ------ | --------------- | ---------- | ---- |
+| i >= 0 | i               |            | from function-references |
+| -0x0b  | `nocont`        |            | new  |
+| -0x18  | `cont`          |            | new  |
+
+### Instructions
```

Comment:
The level of this heading should be increased by one to fit with "Composite types" and "Heap types".
```suggestion
#### Instructions
```